design/spark-co-exist-with-object-storages.md

```diff
@@ -0,0 +1,141 @@
+2. Support any object storage that lakeFS supports.
+3. Do something that adds the least amount of friction to a Spark user as possible.
+
+### Nice to have
+
+Make the solution usable by non-lakeFS users by:
+1. Supporting path translation from and to any type of path.
+2. Make the solution a standalone component with no direct dependency on anything other than Spark itself. 
+
+## Non-goals
+
+1. Convert users from [accessing lakeFS from S3 gateway](../docs/integrations/spark.md#access-lakefs-using-the-s3a-gateway) to [accessing lakeFS with lakeFS-specific Hadoop filesystem](../docs/integrations/spark.md#access-lakefs-using-the-lakefs-specific-hadoop-filesystem).  
+
+## Proposal: Introducing RouterFileSystem
+
+We would like to implement a [HadoopFileSystem](https://github.com/apache/hadoop/blob/2960d83c255a00a549f8809882cd3b73a6266b6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java) 
+that translates object store URIs into lakeFS URIs according to a configurable mapping, and uses a relevant Hadoop file system to 
+perform any file system operation.  
+
+### Handle any interaction with the underlying object store
+
+To allow Spark users to integrate with lakeFS without changing their Spark code, `RouterFileSystem` is configured to be 
+the file system for object store URIs with a certain scheme. For example, the following Spark configurations make 
+`RouterFileSystem` the file system for URIs with `scheme=s3`:
+```shell
+fs.s3.impl=RouterFileSystem
+```
+This will force any file system operation performed on an ibject URI with `scheme=s3` to go through `RouterFileSystem`.   
+
+### URI translation
+
+`RouterFileSystem` has access to a configurable mapping that maps bucket names into lakeFS repositories and prefixes, e.g.
+```yaml
+MapperConfig = {
+    {
+        "source_bucket_pattern": "bucket-a",
+        "mapped_bucket_name": "example-repo",
+        "mapped_prefix": "dev/"
+    }
+}
+```
+
+With the mapping above, the URI `s3://bucket-a/foo.parquet` will be translated into `s3://example-repo/dev/foo.parquet`. 
+
+### Invoke file system operations
+
+After translating URIs to their final form, `RouterFileSystem` will use the translated path and its relevant file system to 
+perform file system operations against the relevant object store. 
+
+Considering the example above, the relevant object store and file system are S3 and `S3AFileSystem` compatibly. `RouterFileSystem` will use `S3AFileSystem` to perform file system operations against S3. e.g. to open `s3://example-repo/dev/foo.parquet`
+
+#### Getting the relevant File System 
+
+Given that `RouterFileSystem` is configured as the file system for URIs with a certain scheme (e.g. `scheme=s3`),
+we need to provide a way to get the relevant file system that can perform actual file system operations against the 
+relevant object store, e.g. `S3AFileSystem`. 
+To do that, we would follow a two-step process: 
+1. add the following Spark configuration:
```

Comment:
Note we will need to use the per bucket configuration or verify that it will be ok to initialize the both filesystems with tsam s3a configurations.
In case the underlying and s3 (for lakefs) will share the same implementation, so they will access the same `fs.s3a.*` configurations.